pkg/pillar/types/global.go

```diff
@@ -114,6 +117,230 @@ func (gs *GlobalStatus) UpdateItemValuesFromGlobalConfig(gc GlobalConfig) {
+
+const (
+	CONFIGITEM_TYPE_INT ConfigItemType = iota + 1
+	CONFIGITEM_TYPE_BOOL
+	CONFIGITEM_TYPE_STRING
+	CONFIGITEM_TYPE_TRISTATE
+)
+
+type ConfigItemSpec struct {
+	Key      string
+	ItemType ConfigItemType
+
+	intMin     uint32
+	intMax     uint32
+	intDefault uint32
+
+	stringDefault   string
+	settingDefault	string
+	boolDefault     bool
+	triStateDefault TriState
+}
+
+type ConfigItemSpecMap struct {
+	itemMap map[string]ConfigItemSpec
+	agentSettings map[string]ConfigItemSpec
+}
+
+func (configMap *ConfigItemSpecMap) AddIntItem(key string, defaultInt uint32, min uint32, max uint32, agentSetting bool) {
+	log.Debugf("Adding int item %s", key)
+	if defaultInt < min || defaultInt > max {
+		log.Fatalf("Adding int item %s failed. Value does not meet given min/max criteria", key)
+	}
+	configItem := ConfigItemSpec{
+		ItemType:   CONFIGITEM_TYPE_INT,
+		Key:        key,
+		intDefault: defaultInt,
+		intMin:     min,
+		intMax:     max,
+	}
+	if agentSetting {
+		configMap.agentSettings[key] = configItem
+	} else {
+		configMap.itemMap[key] = configItem
+	}
+	log.Debugf("Added int item %s", key)
+}
+
+func (configMap *ConfigItemSpecMap) AddBoolItem(key string, defaultBool bool, agentSetting bool) {
+	log.Debugf("Adding bool item %s", key)
+	configItem := ConfigItemSpec{
+		ItemType:    CONFIGITEM_TYPE_BOOL,
+		Key:         key,
+		boolDefault: defaultBool,
+	}
+	if agentSetting {
+		configMap.agentSettings[key] = configItem
+	} else {
+		configMap.itemMap[key] = configItem
+	}
+	log.Debugf("Added bool item %s", key)
+}
+
+func (configMap *ConfigItemSpecMap) AddStringItem(key string, defaultString string, agentSetting bool) {
+	log.Debugf("Adding string item %s", key)
+	configItem := ConfigItemSpec{
+		ItemType:      CONFIGITEM_TYPE_STRING,
+		Key:           key,
+		stringDefault: defaultString,
+	}
+	if agentSetting {
+		configMap.agentSettings[key] = configItem
+	} else {
+		configMap.itemMap[key] = configItem
+	}
+	log.Debugf("Added string item %s", key)
+}
+
+func (configMap *ConfigItemSpecMap) AddTriStateItem(key string, defaultTriState TriState, agentSetting bool) {
+	log.Debugf("Adding tristate item %s", key)
+	configItem := ConfigItemSpec{
+		Key:             key,
+		ItemType:        CONFIGITEM_TYPE_TRISTATE,
+		triStateDefault: defaultTriState,
+	}
+	if agentSetting {
+		configMap.agentSettings[key] = configItem
+	} else {
+		configMap.itemMap[key] = configItem
+	}
+	log.Debugf("Added tristate item %s", key)
+}
+
+type ConfigItem struct {
+	Key      string
+	ItemType ConfigItemType
+
+	intValue	    uint32
+	stringValue		string
+	boolValue	    bool
+	triStateValue	TriState
+}
+
+type ConfigItemMap struct {
+	globalSettings map[string]ConfigItem
+	agentSettings  map[string]ConfigItem
+}
+
+func (specMap *ConfigItemSpecMap) parseItemTestVersion(configMap *ConfigItemMap, item config.ConfigItem) {
+	key := item.Key
+	configItem := ConfigItem {
+		Key:	key,
+	}
+	if specMap.itemMap[key].ItemType == CONFIGITEM_TYPE_INT {
```

Comment:
Here, you are assuming the key is present in the specMap?? How do u know? It may not be present in itemMap?? It can be in AgentSettings as well? Where are you taking care of it?